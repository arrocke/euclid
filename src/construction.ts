import * as Geometry from 'src/geometry'
import {produce} from 'immer'

export interface PointInit extends Geometry.Point {
  type: 'point'
}

export interface LineInit {
  type: 'line'
  left: number
  right: number
}

export interface CircleInit {
  type: 'circle'
  center: number
  edge: number
}

export interface IntersectionInit {
  type: 'intersection'
  one: number
  two: number
  neg: boolean
}

export interface Point extends PointInit, Geometry.Point {
  id: number
}

export interface Line extends Geometry.Line, LineInit {
  id: number
}

export interface Circle extends Geometry.Circle, CircleInit {
  id: number
}

export interface Intersection extends Geometry.Point, IntersectionInit {
  id: number
}

/**
 * Data used to initialize an element in a construction.
 */
export type ElementInit = PointInit | LineInit | CircleInit | IntersectionInit

/**
 * Data used to maintain an element while used within the app.
 */
export type Element = Point | Line | Circle | Intersection

/**
 * Data for a point or intersection in a construction.
 */
export type ConstructionPoint = Point | Intersection

/**
 * Data used to maintain a construction within the app.
 */
export interface Construction {
  elements: Element[]
  points: ConstructionPoint[]
}

/**
 * Describes an error that occurs when manipulating constructions.
 */
export class ConstructionError extends Error {
  public constructor(message?: string) {
    super(message ? `Construction Error: ${message}` : 'Construction Error')
  }
}

/**
 * Returns the list of points updated with intersections with the new element.
 * @param elements the elements to check for intersections with
 * @param points the current list of points in the construction
 * @param newElement the new element being added to the construction
 */
function findIntersections(
  elements: Element[],
  points: ConstructionPoint[],
  newElement: Circle | Line,
): ConstructionPoint[] {
  return produce(points, (draft) => {
    elements.forEach((element) => {
      if (element.type === 'line' || element.type === 'circle') {
        const point1 = Geometry.intersection(element, newElement, true)
        const point2 = Geometry.intersection(element, newElement, false)
        // Ensure the intersection is unique.
        if (draft.every((p) => !Geometry.pointEqual(p, point1))) {
          draft.push({
            ...point1,
            type: 'intersection',
            one: newElement.id,
            two: element.id,
            id: -1,
            neg: true,
          })
        }
        if (draft.every((p) => !Geometry.pointEqual(p, point2))) {
          draft.push({
            ...point2,
            type: 'intersection',
            one: newElement.id,
            two: element.id,
            id: -1,
            neg: false,
          })
        }
      }
    })
  })
}

/**
 * Add a point to a construction
 * @param construction construction to add to
 * @param point point to add
 */
export function addPoint(construction: Construction, point: Geometry.Point): Construction {
  return produce<Construction, Construction>(construction, ({elements, points}) => {
    const {x, y} = point
    const pointElement: Point = {
      x,
      y,
      id: elements.length,
      type: 'point',
    }

    // Prevent adding a duplicate point or intersection.
    if (points.some((p) => Geometry.pointEqual(p, pointElement))) {
      throw new ConstructionError('Point or intersection already exists.')
    }

    // Have to spread so that the objects are different.
    elements.push({...pointElement})
    points.push({...pointElement})
  })
}

/**
 * Add an intersection to a construction.
 * @param construction construction to add to
 * @param oneIndex the index of the first element of the intersection
 * @param twoIndex the index of the second element of the intersection
 * @param neg whether to use the positive or negative intersection if there are two
 */
function addIntersection(
  construction: Construction,
  oneIndex: number,
  twoIndex: number,
  neg: boolean,
): Construction {
  return produce<Construction, Construction>(construction, ({elements, points}) => {
    const one = elements[oneIndex]
    const two = elements[twoIndex]

    // Both elements must exist.
    if (!one || !two) {
      throw new ConstructionError('Elements to construction intersection do not exist.')
    }

    // Both elements must be lines or circles.
    if (
      (one.type !== 'line' && one.type !== 'circle') ||
      (two.type !== 'line' && two.type !== 'circle')
    ) {
      throw new ConstructionError('Elements to construct intersection must be lines or circles')
    }

    const intersection: Intersection = {
      ...Geometry.intersection(one, two),
      neg,
      type: 'intersection',
      one: oneIndex,
      two: twoIndex,
      id: construction.elements.length,
    }

    // Find the intersection in the points list.
    const index = points.findIndex((p) => Geometry.pointEqual(p, intersection))

    // Have to spread so they are different objects.
    points[index] = {...intersection}
    elements.push({...intersection})
  })
}

/**
 * Adds a line to a construction.
 * @param construction construction to add to
 * @param leftIndex the index of one point on the line
 * @param rightIndex the index of another point on the line
 */
export function addLine(
  construction: Construction,
  leftIndex: number,
  rightIndex: number,
): Construction {
  return produce<Construction, Construction>(construction, (draft) => {
    const {elements, points} = draft
    // Find the points for the line.
    const left = points[leftIndex]
    const right = points[rightIndex]

    // Both points must exist.
    if (!left || !right) {
      throw new ConstructionError('Points to contstruct the line do not exist.')
    }

    const computedLine = Geometry.line(left, right)

    // Prevent duplicate lines.
    if (
      elements.some(
        (element) => element.type === 'line' && Geometry.lineEqual(element, computedLine),
      )
    ) {
      throw new ConstructionError('Line already exists.')
    }

    // Add intersections to construction if not already in.
    if (left.id === -1) {
      left.id = elements.length
      elements.push(left)
    }
    if (right.id === -1) {
      right.id = elements.length
      elements.push(right)
    }

    // Construct the line.
    const line: Line = {
      ...computedLine,
      id: elements.length,
      type: 'line',
      left: left.id,
      right: right.id,
    }

    // Add instersections created by the new line.
    draft.points = findIntersections(elements, points, line)
    elements.push(line)
  })
}

/**
 * Adds a circle to a construction
 * @param construction construction to add to
 * @param centerIndex the index of the center point of the circle
 * @param edgeIndex the index of an edge point of the circle
 */
export function addCircle(
  construction: Construction,
  centerIndex: number,
  edgeIndex: number,
): Construction {
  return produce<Construction, Construction>(construction, (draft) => {
    const {elements, points} = draft
    // Find the points for the line.
    const center = points[centerIndex]
    const edge = points[edgeIndex]

    // Both points must exist.
    if (!center || !edge) {
      throw new ConstructionError('Points to contstruct the circle do not exist.')
    }

    const computedCircle = Geometry.circle(center, edge)

    // Prevent duplicate circles.
    if (
      elements.some(
        (element) => element.type === 'circle' && Geometry.circleEqual(element, computedCircle),
      )
    ) {
      throw new ConstructionError('Circle already exists.')
    }

    // Add intersections to construction if not already in.
    if (center.id === -1) {
      center.id = elements.length
      elements.push(center)
    }
    if (edge.id === -1) {
      edge.id = elements.length
      elements.push(edge)
    }

    // Construct the circle.
    const circle: Circle = {
      ...computedCircle,
      id: elements.length,
      type: 'circle',
      center: center.id,
      edge: edge.id,
    }

    // Add instersections created by the new circle.
    draft.points = findIntersections(elements, points, circle)
    elements.push(circle)
  })
}

/**
 * Build a construction from its minimal form
 * @param initialElements array of elements that define the initial construction
 */
export function create(initialElements: ElementInit[] = []): Construction {
  return initialElements.reduce(
    (construction: Construction, element: ElementInit): Construction => {
      switch (element.type) {
        case 'point':
          return addPoint(construction, element)
        case 'line':
          return addLine(construction, element.left, element.right)
        case 'circle':
          return addCircle(construction, element.center, element.edge)
        case 'intersection':
          return addIntersection(construction, element.one, element.two, element.neg)
      }
    },
    {
      elements: [],
      points: [],
    },
  )
}
