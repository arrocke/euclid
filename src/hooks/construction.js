import { useReducer } from 'react'

const { sqrt, pow, abs } = Math
const sq = (n) => pow(n, 2)
const E = 0.00001

const intersection = (e1, e2, neg) => {
  if (e1.type === 'circle' && e2.type === 'line') {
    const temp = e1
    e1 = e2
    e2 = temp
  }

  if (e1.type === 'line' && e2.type === 'line') {
    const x = (e1.b - e2.b) / (e2.m - e1.m)
    const y = e1.m * x + e1.b
    return { x, y }
  } else if (e1.type === 'line' && e2.type === 'circle') {
    const a = 1 + sq(e1.m)
    const b = 2 * (-e2.cx + e1.m * (e1.b - e2.cy))
    const c = sq(e2.cx) + sq(e1.b - e2.cy) - sq(e2.r)

    const x = (-b + (neg ? -1 : 1) * sqrt(sq(b) - 4 * a * c)) / (2 * a)
    const y = e1.m * x + e1.b

    return { x, y }
  } else if (e1.type === 'circle' && e2.type === 'circle') {
    const bl = (sq(e1.cx) - sq(e2.cx) + sq(e1.cy) - sq(e2.cy) - sq(e1.r) + sq(e2.r)) / (-2 * (e2.cy - e1.cy))
    const m = -(e2.cx - e1.cx) / (e2.cy - e1.cy)

    const a = 1 + sq(m)
    const b = 2 * (-e2.cx + m * (bl - e2.cy))
    const c = sq(e2.cx) + sq(bl - e2.cy) - sq(e2.r)

    const x = (-b + (neg ? -1 : 1) * sqrt(sq(b) - 4 * a * c)) / (2 * a)
    const y = m * x + bl

    return { x, y }
  }
}

const calculateIntersections = ({ elements, intersections }, newElement) => {
  const newIntersections = []
  const points = [
    ...elements.filter(x => x.type === 'point' || x.type === 'intersection'),
    ...intersections
  ]
  console.log(points)
  const nonPoints = elements.filter(x => x.type === 'circle' || x.type === 'line')
  for (let element of nonPoints) {

    const pos = {
      ...intersection(newElement, element, false),
      type: 'intersection',
      neg: false
    } 
    const neg = {
      ...intersection(newElement, element, true),
      type: 'intersection',
      neg: true
    } 

    const validPoint = (p) =>
      !isNaN(p.x * p.y) && points.every(i => abs(p.x - i.x) > E || abs(p.y - i.y) > E)

    if (validPoint(pos)) {
      newIntersections.push(pos)
      if (validPoint(neg) && (abs(pos.x - neg.x) > E || abs(pos.y - neg.y) > E)) {
        newIntersections.push(neg)
      }
    } else if (validPoint(neg)) {
      newIntersections.push(neg)
    }
  }
  return newIntersections
}

const findPoint = ({ elements, intersections }, id) => {
  if (typeof id === 'string') {
    return intersections[id.split('-')[0]]
  } else if (!isNaN(id)) {
    return elements[id]
  } else {
    return null
  }
}

const reducer = ({ elements, intersections }, { type, x, y, p1, p2 }) => {
  let id = elements.length
  let point1 = findPoint({ elements, intersections }, p1)
  let point2 = findPoint({ elements, intersections }, p2)

  let newElements = []
  let newIntersections = []

  if (point1 && point1.type === 'intersection' && !point1.id) {
    newElements.push({
      ...point1,
      id: id++
    })
  }
  if (point2 && point2.type === 'intersection' && !point2.id) {
    newElements.push({
      ...point2,
      id: id++
    })
  }

  switch (type) {
    case 'point': {
      newElements.push({ type, x, y, id })
      break
    }
    case 'line': {
      const m = (point2.y - point1.y) / (point2.x - point1.x)
      const b = point1.y - point1.x * m
      const el = { type, p1, p2, m, b, id }
      newElements.push(el)
      newIntersections = calculateIntersections({ elements, intersections, }, el)
      break
    }
    case 'circle': {
      const r = sqrt(sq(point1.x - point2.x) + sq(point1.y - point2.y))
      const el = { type, p1, p2, r, id, cx: point1.x, cy: point1.y }
      newElements.push(el)
      newIntersections = calculateIntersections({ elements, intersections, }, el)
      break
    }
    default: {
      throw new Error(`Cannot add invalid element to construction: ${type}`)
    }
  }

  return {
    elements: [
      ...elements,
      ...newElements
    ],
    intersections: [
      ...intersections.filter(x => x !== point1 && x !== point2),
      ...newIntersections
    ]
  }
}

const useConstruction = () => {
  const [state, dispatch] = useReducer(reducer, {
    elements: [],
    intersections: []
  })

  return [state, dispatch]
}

export default useConstruction