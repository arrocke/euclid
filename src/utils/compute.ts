export interface Point {
  x: number
  y: number
}

export interface Line {
  a: number
  b: number
  c: number
}

export interface Circle {
  h: number
  k: number
  r: number
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isPoint = (obj: any): obj is Point => {
  return typeof obj.x === 'number' && typeof obj.y === 'number'
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isLine = (obj: any): obj is Line => {
  return typeof obj.a === 'number' && typeof obj.b === 'number' && typeof obj.c === 'number'
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isCircle = (obj: any): obj is Circle => {
  return typeof obj.h === 'number' && typeof obj.k === 'number' && typeof obj.r === 'number'
}

export const floatEqual = (a: number, b: number): boolean => {
  return Math.abs(a - b) < 0.0000001
}

export const distance = ({x: x1, y: y1}: Point, {x: x2, y: y2}: Point): number => {
  return Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)
}

export class ComputeError extends Error {
  public constructor(message: string | undefined) {
    super(message ? `ComputeError: ${message}` : 'ComputeError')
  }
}

export const line = (left: Point, right: Point): Line => {
  return {
    a: left.y - right.y,
    b: right.x - left.x,
    c: left.x * right.y - right.x * left.y,
  }
}

export const circle = (center: Point, edge: Point): Circle => {
  return {
    h: center.x,
    k: center.y,
    r: Math.sqrt((center.x - edge.x) ** 2 + (center.y - edge.y) ** 2),
  }
}

export const intersection = (
  el1: Line | Circle,
  el2: Line | Circle,
  neg: boolean = false,
): Point => {
  // Intersection between lines.
  if (isLine(el1) && isLine(el2)) {
    const {a: a1, b: b1, c: c1} = el1
    const {a: a2, b: b2, c: c2} = el2
    if (floatEqual(b1 / a1, b2 / a2)) {
      // Lines overlap completely when one line's coefficients are a multiple of the other.
      if (floatEqual(c1 / a1, c2 / a2)) {
        throw new ComputeError('Lines overlap at all points.')
      }
      // Lines do not intersect when A and B of one line are multiples of the other, but C is not.
      else {
        throw new ComputeError('Lines do not intersect.')
      }
    }
    const x = (c2 * b1 - c1 * b2) / (a1 * b2 - a2 * b1)
    const y = (c1 * a2 - c2 * a1) / (a1 * b2 - a2 * b1)
    return {x, y}
  }
  // Intersection between circles.
  else if (isCircle(el1) && isCircle(el2)) {
    const {h: h1, k: k1, r: r1} = el1
    const {h: h2, k: k2, r: r2} = el2
    // Circles overlap completely when their centers and radii are the same.
    if (floatEqual(h1, h2) && floatEqual(k1, k2) && floatEqual(r1, r2)) {
      throw new ComputeError('Circles overlap at all points.')
    }
    const d = distance({x: h1, y: k1}, {x: h2, y: k2})
    // Circles do not overlap when the disance between centers is greater than the sum of the radii.
    if (d > r1 + r2) {
      throw new ComputeError('Circles do not intersect.')
    }
    const a = (r1 ** 2 - r2 ** 2 + d ** 2) / (2 * d)
    const h = Math.sqrt(r1 ** 2 - a ** 2)
    const cx = h1 + (a * (h2 - h1)) / d
    const cy = k1 + (a * (k2 - k1)) / d
    const x = cx + ((neg ? 1 : -1) * h * (k2 - k1)) / d
    const y = cy + ((neg ? -1 : 1) * h * (h2 - h1)) / d
    return {x, y}
  }
  // Intersection between a line and circle.
  else {
    // Ensure that the first element is a line and the second a circle.
    if (isCircle(el1)) {
      const temp = el1
      el1 = el2
      el2 = temp
    }
    const {a, b, c} = el1 as Line
    const {h, k, r} = el2 as Circle
    const d = Math.sqrt(r ** 2 - (a * h + b * k + c) ** 2 / (a ** 2 + b ** 2))
    const xc = (b * (b * h - a * k) - a * c) / (a ** 2 + b ** 2)
    const yc = (a * (-b * h + a * k) - b * c) / (a ** 2 + b ** 2)
    const x = xc + (neg ? -1 : 1) * b * Math.sqrt(d ** 2 / (a ** 2 + b ** 2))
    const y = yc + (neg ? 1 : -1) * a * Math.sqrt(d ** 2 / (a ** 2 + b ** 2))
    if (isNaN(x) || isNaN(y)) {
      throw new ComputeError('Circle and line do not intersect.')
    }
    return {x, y}
  }
}
