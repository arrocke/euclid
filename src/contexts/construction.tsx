import React, {createContext, useReducer, useEffect, useContext, useMemo} from 'react'
import * as compute from 'src/utils/compute'
import {ConstructionState, ElementState, Element} from 'src/types'

interface ConstructionContextValue {
  construction: ConstructionState
  dispatch: React.Dispatch<ReducerAction>
}

interface ConstructionProviderProps {
  value?: ConstructionContextValue
  initial: Element[] | Promise<Element[]>
}

interface InitAction {
  type: 'init'
  data: Element[]
}

interface PointAction {
  type: 'point'
  data: {x: number; y: number}
}

type ReducerAction = InitAction | PointAction

const ConstructionContext = createContext<ConstructionContextValue | null>(null)

const computeIntersection = (
  el1: ElementState,
  el2: ElementState,
  neg: boolean = false,
): compute.Point | null => {
  if (el1.type === 'c' && el2.type === 'c') {
    return compute.circleIntersection(el1, el2, neg)
  } else if (el1.type === 'l' && el2.type === 'l') {
    return compute.lineIntersection(el1, el2)
  } else if (el1.type === 'l' && el2.type === 'c') {
    return compute.circleLineIntersection(el2, el1, neg)
  } else if (el1.type === 'c' && el2.type === 'l') {
    return compute.circleLineIntersection(el1, el2, neg)
  } else {
    return null
  }
}

const reducer: React.Reducer<ConstructionState, ReducerAction> = (state, action) => {
  const EPSILON = 0.00001
  switch (action.type) {
    case 'init': {
      return action.data.reduce<ConstructionState>(
        ({elements, points}, el) => {
          // Add points to the list by finding intersection with previous elements.
          const addPoints = (element: ElementState) => {
            elements.forEach((el) => {
              const point1 = computeIntersection(element, el, true)
              if (
                point1 &&
                points.every(
                  ({x, y}) => Math.abs(x - point1.x) > EPSILON || Math.abs(y - point1.y) > EPSILON,
                )
              ) {
                points.push(point1)
              }
              const point2 = computeIntersection(element, el, false)
              if (
                point2 &&
                points.every(
                  ({x, y}) => Math.abs(x - point2.x) > EPSILON || Math.abs(y - point2.y) > EPSILON,
                )
              ) {
                points.push(point2)
              }
            })
          }

          switch (el.type) {
            case 'p': {
              points.push({...el, id: elements.length})
              elements.push(el)
              break
            }
            case 'l': {
              const left = elements[el.left]
              const right = elements[el.right]
              if (['i', 'p'].includes(left.type) && ['i', 'p'].includes(right.type)) {
                const line = {
                  ...el,
                  ...compute.line(left as compute.Point, right as compute.Point),
                }
                addPoints(line)
                elements.push(line)
              } else {
                throw new Error('Invalid line')
              }
              break
            }
            case 'c': {
              const center = elements[el.center]
              const edge = elements[el.edge]
              if (['i', 'p'].includes(center.type) && ['i', 'p'].includes(edge.type)) {
                const circle = {
                  ...el,
                  ...compute.circle(center as compute.Point, edge as compute.Point),
                }
                addPoints(circle)
                elements.push(circle)
              } else {
                throw new Error('Invalid circle')
              }
              break
            }
            case 'i': {
              const el1 = elements[el.element1]
              const el2 = elements[el.element2]
              const intersection = computeIntersection(el1, el2, el.neg)
              if (intersection) {
                const i = points.findIndex(
                  (point) =>
                    Math.abs(intersection.x - point.x) < EPSILON &&
                    Math.abs(intersection.y - point.y) < EPSILON,
                )
                points[i].id = elements.length
                elements.push({
                  ...el,
                  ...intersection,
                })
              } else {
                throw new Error('Invalid intersection')
              }
              break
            }
          }
          return {elements, points}
        },
        {elements: [], points: []},
      )
    }
    case 'point': {
      return {
        elements: [...state.elements, {type: 'p', x: action.data.x, y: action.data.y}],
        points: [...state.points, {id: state.elements.length, x: action.data.x, y: action.data.y}],
      }
    }
  }
  return state
}

export const ConstructionProvider: React.FC<ConstructionProviderProps> = ({
  value,
  initial,
  children,
}) => {
  const [construction, dispatch] = useReducer(reducer, {elements: [], points: []})

  // Reloads the construction when the initial prop changes.
  useEffect(() => {
    const effect = async () =>
      dispatch({type: 'init', data: initial instanceof Promise ? await initial : initial})
    effect()
  }, [initial])

  const contextValue = useMemo(() => value || {construction, dispatch}, [construction])

  return (
    <ConstructionContext.Provider value={contextValue}>{children}</ConstructionContext.Provider>
  )
}

export const useConstruction = () => {
  const context = useContext(ConstructionContext)
  if (!context) {
    throw new Error('useConstruction must be used within a ConstructionProvider.')
  }
  const {
    construction: {elements, points},
    dispatch,
  } = context
  const addPoint = (x: number, y: number): void => {
    dispatch({
      type: 'point',
      data: {x, y},
    })
  }
  return {elements, points, addPoint}
}
