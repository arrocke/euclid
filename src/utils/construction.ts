import * as Compute from 'src/utils/compute'

export interface PointInit extends Compute.Point {
  type: 'point'
}

export interface LineInit {
  type: 'line'
  left: number
  right: number
}

export interface CircleInit {
  type: 'circle'
  center: number
  edge: number
}

export interface IntersectionInit {
  type: 'intersection'
  one: number
  two: number
  neg: boolean
}

export interface Point extends PointInit, Compute.Point {
  id: number
}

export interface Line extends Compute.Line, LineInit {
  id: number
}

export interface Circle extends Compute.Circle, CircleInit {
  id: number
}

export interface Intersection extends Compute.Point, IntersectionInit {
  id: number
}

export interface ConstructionIntersection extends Compute.Point, IntersectionInit {}

export type ElementInit = PointInit | LineInit | CircleInit | IntersectionInit
export type Element = Point | Line | Circle | Intersection
export type ConstructionPoint = Point | Intersection

export interface Construction {
  elements: Element[]
  points: ConstructionPoint[]
}

function addIntersections(
  elements: Element[],
  points: ConstructionPoint[],
  newElement: Circle | Line,
) {
  elements.forEach((element) => {
    if (element.type === 'line' || element.type === 'circle') {
      const point1 = Compute.intersection(element, newElement, true)
      const point2 = Compute.intersection(element, newElement, false)
      // Ensure the intersection is unique.
      if (
        points.every(
          ({x, y}) => !Compute.floatEqual(x, point1.x) || !Compute.floatEqual(y, point1.y),
        )
      ) {
        points.push({
          ...point1,
          type: 'intersection',
          one: newElement.id,
          two: element.id,
          id: -1,
          neg: true,
        })
      }
      if (
        points.every(
          ({x, y}) => !Compute.floatEqual(x, point2.x) || !Compute.floatEqual(y, point2.y),
        )
      ) {
        points.push({
          ...point2,
          type: 'intersection',
          one: newElement.id,
          two: element.id,
          id: -1,
          neg: false,
        })
      }
    }
  })
}

export function create(elementInits: ElementInit[] = []): Construction {
  const elements: Element[] = []
  const points: ConstructionPoint[] = []
  elementInits.forEach((element) => {
    switch (element.type) {
      case 'point': {
        const point = {
          ...element,
          id: elements.length,
        }
        elements.push(point)
        points.push(point)
        break
      }
      case 'line': {
        const line = {
          ...element,
          ...Compute.line(elements[element.left] as Point, elements[element.right] as Point),
          id: elements.length,
        }
        addIntersections(elements, points, line)
        elements.push(line)
        break
      }
      case 'circle': {
        const circle = {
          ...element,
          ...Compute.circle(elements[element.center] as Point, elements[element.edge] as Point),
          id: elements.length,
        }
        addIntersections(elements, points, circle)
        elements.push(circle)
        break
      }
      case 'intersection': {
        const one = elements[element.one]
        const two = elements[element.two]
        if (
          (one.type === 'line' || one.type === 'circle') &&
          (two.type === 'line' || two.type === 'circle')
        ) {
          const intersection = {
            ...element,
            ...Compute.intersection(one, two),
            id: elements.length,
          }
          const point = points.filter(
            ({x, y}) =>
              Compute.floatEqual(x, intersection.x) && Compute.floatEqual(y, intersection.y),
          )[0]
          const index = points.indexOf(point)
          points.splice(index, 1, intersection)
          elements.push(intersection)
          break
        }
      }
    }
  })
  return {elements, points}
}
