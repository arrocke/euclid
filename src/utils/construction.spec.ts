import * as Construction from 'src/utils/construction'

describe('create', () => {
  test('when called with no arguments, returns an empty construction', () => {
    expect(Construction.create()).toMatchSnapshot()
  })

  test('initializes construction from argument', () => {
    expect(
      Construction.create([
        {type: 'point', x: 1, y: 2},
        {type: 'point', x: -5, y: -2},
        {type: 'line', left: 0, right: 1},
        {type: 'circle', center: 0, edge: 1},
        {type: 'circle', center: 1, edge: 0},
        {type: 'intersection', one: 3, two: 4, neg: true},
        {type: 'line', left: 0, right: 5},
        {type: 'line', left: 1, right: 5},
      ]),
    ).toMatchSnapshot()
  })
})
