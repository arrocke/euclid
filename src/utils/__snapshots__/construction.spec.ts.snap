// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create initializes construction from argument 1`] = `
Object {
  "elements": Array [
    Object {
      "id": 0,
      "type": "point",
      "x": 1,
      "y": 2,
    },
    Object {
      "id": 1,
      "type": "point",
      "x": -5,
      "y": -2,
    },
    Object {
      "a": 4,
      "b": -6,
      "c": 8,
      "id": 2,
      "left": 0,
      "right": 1,
      "type": "line",
    },
    Object {
      "center": 0,
      "edge": 1,
      "h": 1,
      "id": 3,
      "k": 2,
      "r": 7.211102550927978,
      "type": "circle",
    },
    Object {
      "center": 1,
      "edge": 0,
      "h": -5,
      "id": 4,
      "k": -2,
      "r": 7.211102550927978,
      "type": "circle",
    },
    Object {
      "id": 5,
      "neg": true,
      "one": 3,
      "two": 4,
      "type": "intersection",
      "x": 1.4641016151377544,
      "y": -5.196152422706632,
    },
    Object {
      "a": 7.196152422706632,
      "b": 0.4641016151377544,
      "c": -8.12435565298214,
      "id": 6,
      "left": 0,
      "right": 5,
      "type": "line",
    },
    Object {
      "a": 3.196152422706632,
      "b": 6.464101615137754,
      "c": 28.908965343808667,
      "id": 7,
      "left": 1,
      "right": 5,
      "type": "line",
    },
  ],
  "points": Array [
    Object {
      "id": 0,
      "type": "point",
      "x": 1,
      "y": 2,
    },
    Object {
      "id": 1,
      "type": "point",
      "x": -5,
      "y": -2,
    },
    Object {
      "id": -1,
      "neg": true,
      "one": 3,
      "two": 2,
      "type": "intersection",
      "x": 6.999999999999999,
      "y": 6,
    },
    Object {
      "id": -1,
      "neg": false,
      "one": 4,
      "two": 2,
      "type": "intersection",
      "x": -11,
      "y": -6,
    },
    Object {
      "id": -1,
      "neg": true,
      "one": 4,
      "two": 3,
      "type": "intersection",
      "x": -5.464101615137754,
      "y": 5.196152422706632,
    },
    Object {
      "id": 5,
      "neg": true,
      "one": 3,
      "two": 4,
      "type": "intersection",
      "x": 1.4641016151377544,
      "y": -5.196152422706632,
    },
    Object {
      "id": -1,
      "neg": true,
      "one": 6,
      "two": 3,
      "type": "intersection",
      "x": 0.5358983848622458,
      "y": 9.196152422706632,
    },
    Object {
      "id": -1,
      "neg": true,
      "one": 7,
      "two": 4,
      "type": "intersection",
      "x": -11.464101615137753,
      "y": 1.1961524227066316,
    },
  ],
}
`;

exports[`create when called with no arguments, returns an empty construction 1`] = `
Object {
  "elements": Array [],
  "points": Array [],
}
`;
